lambdaurl/http_handler.go

```diff
@@ -0,0 +1,103 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+// Package lambdaurl serves requests from Lambda Function URLs using http.Handler.
+package lambdaurl
+
+import (
+	"context"
+	"encoding/base64"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-lambda-go/events"
+	"github.com/aws/aws-lambda-go/lambda"
+)
+
+type httpResponseWriter struct {
+	header http.Header
+	writer io.Writer
+	once   sync.Once
+	status chan<- int
+}
+
+func (w *httpResponseWriter) Header() http.Header {
+	return w.header
+}
+
+func (w *httpResponseWriter) Write(p []byte) (int, error) {
+	w.once.Do(func() { w.status <- http.StatusOK })
+	return w.writer.Write(p)
+}
+
+func (w *httpResponseWriter) WriteHeader(statusCode int) {
+	w.once.Do(func() { w.status <- statusCode })
+}
+
+type requestContextKey struct{}
+
+// RequestFromContext returns the *events.LambdaFunctionURLRequest from a context.
+func RequestFromContext(ctx context.Context) (*events.LambdaFunctionURLRequest, bool) {
+	req, ok := ctx.Value(requestContextKey{}).(*events.LambdaFunctionURLRequest)
+	return req, ok
+}
+
+// Wrap converts an http.Handler into a lambda request handler.
+// Only Lambda Function URLs configured with `InvokeMode: RESPONSE_STREAM` are supported with the returned handler.
+// The response body of the handler will conform the the content-type `application/vnd.awslambda.http-integration-response`
+func Wrap(handler http.Handler) func(context.Context, *events.LambdaFunctionURLRequest) (*events.LambdaFunctionURLStreamingResponse, error) {
+	return func(ctx context.Context, request *events.LambdaFunctionURLRequest) (*events.LambdaFunctionURLStreamingResponse, error) {
+		var body io.Reader = strings.NewReader(request.Body)
+		if request.IsBase64Encoded {
+			body = base64.NewDecoder(base64.StdEncoding, body)
+		}
+		url := "https://" + request.RequestContext.DomainName + request.RawPath
+		if request.RawQueryString != "" {
+			url += "?" + request.RawQueryString
+		}
+		ctx = context.WithValue(ctx, requestContextKey{}, request)
+		httpRequest, err := http.NewRequestWithContext(ctx, request.RequestContext.HTTP.Method, url, body)
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range request.Headers {
+			httpRequest.Header.Add(k, v)
+		}
+		status := make(chan int) // Signals when it's OK to start returning the response body to Lambda
+		header := http.Header{}
+		r, w := io.Pipe()
+		go func() {
+			defer close(status)
+			defer w.Close() // TODO: recover and CloseWithError the any panic value once the runtime API client supports plumbing fatal errors through the reader
```

Comment:
https://github.com/aws/aws-lambda-go/issues/504

Reply:
Will get back to this in the follow-up for all returned writers. As for now allowing the panic to crash the process and log is the "most correct" thing do do, at least compared to swallowing it without any logging / reporting!